storage/remote/queue_manager.go

```diff
@@ -302,11 +341,96 @@ func NewQueueManager(reg prometheus.Registerer, metrics *queueManagerMetrics, lo
 	}
 
 	t.watcher = wal.NewWatcher(reg, wal.NewWatcherMetrics(reg), logger, client.Name(), t, walDir)
+	if t.cfg.SendMetadata {
+		t.metadataWatcher = NewMetadataWatcher(logger, sm, client.Name(), t, t.cfg.MetadataSendInterval, flushDeadline)
+	}
 	t.shards = t.newShards()
 
 	return t
 }
 
+// AppendMetadata queues metadata to be sent to the remote storage. Metadata is sent all at once and is not parallelized.
+func (t *QueueManager) AppendMetadata(ctx context.Context, metadata []scrape.MetricMetadata) {
+	metadataSet := map[scrape.MetricMetadata]struct{}{}
+
+	// We create a set of the metadata by deduplicating based on the attributes of a
+	// scrape.MetricMetadata. In this case, a combination of metric name, help, type, and unit.
+	for _, entry := range metadata {
+		metadataSet[entry] = struct{}{}
+	}
+
+	mm := make([]prompb.MetricMetadata, 0, len(metadataSet))
+	for entry := range metadataSet {
+		mm = append(mm, prompb.MetricMetadata{
+			MetricName: entry.Metric,
+			Help:       entry.Help,
+			Type:       metricTypeToMetricTypeProto(entry.Type),
+			Unit:       entry.Unit,
+		})
+	}
+
+	err := t.sendMetadataWithBackoff(ctx, mm)
+
+	if err != nil {
+		level.Error(t.logger).Log("msg", "non-recoverable error while sending metadata", "count", len(metadata), "err", err)
+	}
+}
+
+func (t *QueueManager) sendWriteRequestWithBackoff(ctx context.Context, req []byte, reportFailureFunc func()) (elapsed float64, err error) {
+	backoff := t.cfg.MinBackoff
+
+	for {
+		select {
+		case <-ctx.Done():
+			return -1, ctx.Err()
+		default:
+		}
+
+		begin := time.Now()
+		err := t.client.Store(context.Background(), req)
+		elapsed = time.Since(begin).Seconds()
+
+		if err == nil {
+			return elapsed, nil
+		}
+
+		if _, ok := err.(recoverableError); !ok {
+			return -1, err
+		}
+
+		if reportFailureFunc != nil {
+			reportFailureFunc()
```

Comment:
Why it's optional? Number of retries would be nice to know as metric I guess.

Reply:
My comment above explains that.

Reply:
This is no longer the case 😄